atc/exec/put_inputs.go

```diff
@@ -58,14 +60,27 @@ func (i specificInputs) FindAll(artifacts *build.Repository) (map[string]runtime
 
 	inputs := map[string]runtime.Artifact{}
 
-	for _, i := range i.inputs {
-		artifact, found := artifactsMap[build.ArtifactName(i)]
+	for _, input := range i.inputs {
+		if input == atc.InputsDetect {
+			detectedInputs, err := NewDetectInputs(i.params).FindAll(artifacts)
```

Comment:
What's the motivation for structuring it like this in the code, with `detect` being stored in `specificInputs`? What happens if a user has an input called `detect`? I realize this doesn't change how users need to write their pipelines, but if a user has:
```yaml
- put: resource
  inputs:
    - foo
    - bar
    - detect
```
and the input is actually called `detect` then won't we end up trying to "detect" inputs when we shouldn't?

Also feels odd to have `specificInputs` doing two things now: 1) finding specific inputs and 2) detecting inputs. I think this overloads what `specificInputs` does. Is there a good reason to do this? I don't see why we have to insert `detect` into the `inputs` array.

Reply:
@taylorsilva The motivation is from the third error I described above. Say, a `put` depends on 10 inputs (9 can be detected, 1 is implicit). After a user set `inputs: detect` to the `put`, it hits an error of "some folder not found" because the implicit input cannot be detected. Then the user can simply change to `inputs: [detect, implicit-one]`, instead of having to list all rest 9 inputs.

If a `put` step happens to have an input named "detect", both `inputs: detect` and `inputs: [detect, other-input]` would still work. Only cost of `inputs: [detect, other-input]` is parsing `params` which suppose to be a tiny effort, I won't worry about that.

Embedding `detect` into `specificInputs` is just an implementation detail, it doesn't impact user experience. This implementation avoids duplicate code if we want to support embedding "detect" in inputs array.

I also realize one thing that I didn't realize yesterday. Only problem of embedding "detect" in inputs array is, if a `put` happens to have an implicit input named "detect", that must be explicitly specified in inputs array, then my current design will not be able to support that. To solve that problem, maybe we can use an escape character, `inputs: [detect, \detect]`, where the first "detect" means to auto detect inputs from `params`, and the second "\detect" means the input named "detect" should be included. WDYT?

Reply:
This is when I wish I had a magic wand and could have concourse expose `inputs` as:
```yaml
- put: resource
    inputs:
      detect: true
      specific:
      - foo
      - bar
```
or something analogous to that.

The solution you proposed would work but it would be hacky and even documenting it feels awkward. https://concourse-ci.org/jobs.html#schema.step.put-step.inputs

I get that this commit gets "detect" closer to being a 100% solution for you, but I really don't like what it's doing to the code and the UX it gives users writing pipelines.

Currently, without this PR, the user and you as operator don't have zero options right now. Specifying specific inputs seems perfectly valid here, given some inputs are implicit. If the inputs do change the pipeline author will know on the first build with the change, given they don't catch it themselves. And if inputs do change they probably have to update the `params` anyways, right beside where the `inputs` are.

Using detect is convenient, you don't have to change the pipeline as much in the future. Though when I think about it, what's the workflow for someone using detect and changing the name of their inputs? They have to change the params anyways! They'll be modifying the step already. 

Right now I think adding `detect` as a keyword to the list of inputs would be confusing. It is neat but it complicates the code and people's understanding of pipelines. It won't be obvious on a first read if the `- detect` item is actually an input called detect or is it "detect"ing inputs. Someone will have to go into the docs or leave a comment in their pipeline about the special "detect" item. I'm not in favour of the change from this specific commit. You already need to advise your users to use `detect`, I don't think it's that much more of a stretch to tell them: "try detect first and then list all inputs if detect doesn't catch everything".

Reply:
@taylorsilva If that prevents this PR from proceeding, I comprise. Let me revert the change of adding `detect` to inputs list.